draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,32 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's target
+and port override the alt-authority.  For example, suppose that
+"https://example.com" sends an Alt-Svc field value of:
+
+    Alt-Svc: h2="alt.example.com:443", h3=":8443"
+
+The client would retrieve the following HTTPS records:
+    alt.example.com. IN HTTPS 1 . alpn=h2,h3 ech=...
+    _8443._https.example.com. IN HTTPS 1 alt3.example.com. (
+        port=9443 alpn=h2,h3 ech=... )
+
+The client could then attempt an HTTP/3 connection to `alt3.example.com:9443`
+with ECH, or an HTTP over TLS connection to `alt.example.com:443` with ECH,
+as these options are consistent with both an Alt-Svc field value and its
+HTTPS record.  This specification does not alter or clarify the
+interpretation of Alt-Svc's `protocol-id` field.
+
+Origins that publish an "ech" SvcParam in their HTTPS record SHOULD
+also publish an "ech" SvcParam for any alt-authorities.  Otherwise,
+clients may reveal the unencrypted ClientHello during Alt-Svc connections.
```

Comment:
I would prefer:

> Clients that connect to an alt-authority would reveal the name of the server in an unencrypted ClientHello.  Clients SHOULD NOT use an alt-authority if the original connection uses ECH and no "ech" SvcParam is provided for the alt-authority.

This is not saying that the client could (or may) reveal identity, but saying that if it did connect then it would.  THEN I want a recommendation against doing the stupid thing.  I would settle for just the first sentence, but I do really prefer both.

Reply:
Just to be clear, you're proposing that implementations should add an indication to each Alt-Svc cache entry about whether ECH was in use on the connection over which that entry was received, so that ECH consistency can be enforced when that cache entry is used (perhaps several weeks later)?

Reply:
No, I'm suggesting that client implementations that start using ECH keep using ECH.  The decision to use an alternative is made in the moment and can be based on what is active. That's consistent with the other text you have.  

I hadn't fully considered the point about caching.  That is, if we use ECH on Tuesday, but fetch and reject an Alt-Svc advertisement for lack of ECH, do you cache that entry?  

This is really something that an Alt-Svc revision needs to deal with, so we might need to be careful here.  I can imagine that the decision about whether to insist on ECH would be made anew on Friday.  My preference is that yes, if you do cache the Alt-Svc advertisement you also remember that it was disqualified for lack of ECH or just that it was not acceptable at the time it was retrieved.  For me, that marker would be cause to avoid connecting immediately to the alternative without first going to the primary.  Which gets back to the SHOULD.

Reply:
I certainly hope most clients would get smart about not changing away from an ECH connection to a non-ECH connection, but I think this is hard to create any actual rules about without over-specifying things and I suggest we not get into it.  E.g., how long after a successful ECH connection should a client be rejecting non-ECH Alt-Svc?

Plus, I think leaving this up to the client to make a good decision is consistent with the rest of the spec.  If the server gives some ECH endpoints and some non-ECH endpoints, it's up to the client to make the decisions over how/if to prefer endpoints that support ECH.  So the spec is saying if a server really wants to require ECH for ECH-capable clients, they need to include ECH on all endpoints, including Alt-Svc endpoints, and this is the recommended/should'ed server configuration.  If the server doesn't configure that, most clients will ideally still choose to prefer ECH, but that's up to the client, and some clients may have reason to prefer any route to h3, thus potentially disabling ECH.

Reply:
> My preference is that yes, if you do cache the Alt-Svc advertisement you also remember that it was disqualified for lack of ECH or just that it was not acceptable at the time it was retrieved.

Alt-Svc field values don't say anything about ECH.  That's a property of the HTTPS record for the alt-authority, which is not known (and may change) until the client attempts the Alt-Svc connection, which may not be until much later.

It sounds like we want to support this behavior, but the details are too unsettled to make a normative recommendation.  I've adjusted the explanation based on @martinthomson's suggestion, and added the following sentence:
> Clients MAY impose their own consistency conditions on Alt-Svc connections, e.g. requiring ECH if the origin's SvcParams offered ECH.

Please review.

Reply:
Current text still looks good to me.  The important thing here is that we give clients a technical path to ECH and make clear to domains the strongly recommended way to configure it to make sure clients can use it.  I think we have that.